File: frontend/web/services/resume.service.ts
import { api } from '@/lib/api-client';
import type {
  SavedResume,
  CreateResumeRequest,
  UpdateResumeRequest,
  ResumePreviewData,
  TemplateInfo,
  ResumeFileMetadata,
} from '@/types/resume.types';

/**
 * Get all user data for resume preview (JSON format)
 */
export async function getResumePreviewData(): Promise<ResumePreviewData> {
  const response = await api.get<ResumePreviewData>('/resume/preview');
  return response.data;
}

/**
 * Get available resume templates
 */
export async function getTemplates(): Promise<TemplateInfo[]> {
  const response = await api.get<TemplateInfo[]>('/resume/templates');
  return response.data;
}

/**
 * Get all saved resumes for the authenticated user
 */
export async function getSavedResumes(): Promise<SavedResume[]> {
  const response = await api.get<{ data: SavedResume[] }>('/resume/saved');
  return response.data.data;
}

/**
 * Get a specific saved resume by ID
 */
export async function getSavedResume(id: string): Promise<SavedResume> {
  const response = await api.get<SavedResume>(`/resume/saved/${id}`);
  return response.data;
}

/**
 * Create a new saved resume
 */
export async function createSavedResume(data: CreateResumeRequest): Promise<SavedResume> {
  const response = await api.post<SavedResume>('/resume/saved', data);
  return response.data;
}

/**
 * Update a saved resume
 */
export async function updateSavedResume(
  id: string,
  data: UpdateResumeRequest
): Promise<SavedResume> {
  const response = await api.patch<SavedResume>(`/resume/saved/${id}`, data);
  return response.data;
}

/**
 * Delete a saved resume (soft delete)
 */
export async function deleteSavedResume(id: string): Promise<void> {
  await api.delete(`/resume/saved/${id}`);
}

/**
 * Generate and download a PDF resume
 * Opens the PDF in a new window/tab
 */
export async function generateAndDownloadResume(): Promise<void> {
  const response = await api.post('/resume/generate', {}, { responseType: 'blob' });

  // Create a blob URL and trigger download
  const blob = new Blob([response.data], { type: 'application/pdf' });
  const url = window.URL.createObjectURL(blob);
  const link = document.createElement('a');
  link.href = url;
  link.download = `resume-${new Date().toISOString().split('T')[0]}.pdf`;
  document.body.appendChild(link);
  link.click();
  document.body.removeChild(link);
  window.URL.revokeObjectURL(url);
}

/**
 * Get download URL for a stored resume PDF
 */
export async function getResumeDownloadUrl(resumeId: string): Promise<{ url: string }> {
  const response = await api.get<{ url: string }>(`/resume/download/${resumeId}`);
  return response.data;
}

/**
 * Get metadata about stored resume PDF file
 */
export async function getResumeFileInfo(resumeId: string): Promise<ResumeFileMetadata> {
  const response = await api.get<ResumeFileMetadata>(`/resume/file-info/${resumeId}`);
  return response.data;
}

/**
 * Delete stored resume PDF file from storage
 */
export async function deleteResumeFile(resumeId: string): Promise<void> {
  await api.delete(`/resume/file/${resumeId}`);
}

/**
 * Download resume PDF with a given resumeId
 */
export async function downloadResumePdf(resumeId: string): Promise<void> {
  const { url } = await getResumeDownloadUrl(resumeId);
  window.open(url, '_blank');
}

File: frontend/app/lib/features/resume/services/resume_service.dart
import 'dart:typed_data';

import 'package:dio/dio.dart';
import 'package:flutter_riverpod/flutter_riverpod.dart';

import '../../../core/network/api_client.dart';
import '../../../core/utils/logger.dart';
import '../models/resume_model.dart';

final resumeServiceProvider = Provider<ResumeService>((ref) {
  return ResumeService(ref.read(apiClientProvider));
});

class ResumeService {
  final ApiClient _api;

  ResumeService(this._api);

  /// Get all user data for resume preview
  Future<ResumePreviewData> getResumePreviewData() async {
    try {
      final response = await _api.get('/api/resume/preview');
      return ResumePreviewData.fromJson(response.data);
    } catch (e) {
      logger.e('Failed to fetch resume preview data', error: e);
      rethrow;
    }
  }

  /// Get available resume templates
  Future<List<TemplateInfo>> getTemplates() async {
    try {
      final response = await _api.get('/api/resume/templates');
      return (response.data as List)
          .map((e) => TemplateInfo.fromJson(e))
          .toList();
    } catch (e) {
      logger.e('Failed to fetch templates', error: e);
      rethrow;
    }
  }

  /// Get all saved resumes for the authenticated user
  Future<List<SavedResume>> getSavedResumes() async {
    try {
      final response = await _api.get('/api/resume/saved');

      final data = response.data;
      final List list = (data is Map && data.containsKey('data'))
          ? data['data']
          : data;

      return list.map((e) => SavedResume.fromJson(e)).toList();
    } catch (e) {
      logger.e('Failed to fetch saved resumes', error: e);
      rethrow;
    }
  }

  /// Get a specific saved resume by ID
  Future<SavedResume> getSavedResume(String id) async {
    try {
      final response = await _api.get('/api/resume/saved/$id');
      return SavedResume.fromJson(response.data);
    } catch (e) {
      logger.e('Failed to fetch saved resume $id', error: e);
      rethrow;
    }
  }

  /// Create a new saved resume
  Future<SavedResume> createSavedResume(CreateResumeRequest request) async {
    try {
      final response = await _api.post(
        '/api/resume/saved',
        data: request.toJson(),
      );
      return SavedResume.fromJson(response.data);
    } catch (e) {
      logger.e('Failed to create saved resume', error: e);
      rethrow;
    }
  }

  /// Update a saved resume
  Future<SavedResume> updateSavedResume(
    String id,
    UpdateResumeRequest request,
  ) async {
    try {
      final response = await _api.patch(
        '/api/resume/saved/$id',
        data: request.toJson(),
      );
      return SavedResume.fromJson(response.data);
    } catch (e) {
      logger.e('Failed to update saved resume $id', error: e);
      rethrow;
    }
  }

  /// Delete a saved resume
  Future<void> deleteSavedResume(String id) async {
    try {
      await _api.delete('/api/resume/saved/$id');
    } catch (e) {
      logger.e('Failed to delete saved resume $id', error: e);
      rethrow;
    }
  }

  /// Generate a PDF resume for the authenticated user
  /// Returns the PDF file as bytes
  Future<Uint8List> generateMyResume() async {
    try {
      final response = await _api.client.post(
        '/api/resume/generate',
        options: Options(
          responseType: ResponseType.bytes,
        ),
      );
      return Uint8List.fromList(response.data);
    } catch (e) {
      logger.e('Failed to generate resume PDF', error: e);
      rethrow;
    }
  }
}
